packages/next/src/server/base-server.ts

```diff
@@ -2769,6 +2770,23 @@ export default abstract class Server<ServerOptions extends Options = Options> {
         )
       }
 
+      if (cachedData.headers) {
+        const resHeaders = cachedData.headers
+        for (const key of Object.keys(resHeaders)) {
+          if (key === NEXT_CACHE_TAGS_HEADER) {
+            // Not sure if needs to be special.
+            continue
+          }
+          let v = resHeaders[key]
+          if (typeof v !== 'undefined') {
+            if (typeof v === 'number') {
+              v = v.toString()
+            }
+            res.setHeader(key, v)
+          }
```

Comment:
Personally prefer to perform an early `continue/return` rather than add more nesting, but not required!